tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedXConnection.java

```diff
@@ -165,17 +162,26 @@ protected CloseableIteration<? extends BindingSet, QueryEvaluationException> eva
 				});
 				queryBindings = actualQueryBindings;
 			}
-			CloseableIteration<? extends BindingSet, QueryEvaluationException> res = strategy.evaluate(query,
-					queryBindings);
-
-			// mark the query as PassedThrough, such that outer result handlers are aware of this
-			// Note: for SingleSourceQuery (i.e. where we use pass through) res is explicitly
-			// EmptyIteration. Thus we can use it as indicator
-			if (_orgQuery instanceof PassThroughTupleExpr && res instanceof EmptyIteration) {
-				((PassThroughTupleExpr) _orgQuery).setPassedThrough(true);
+
+			CloseableIteration<? extends BindingSet, QueryEvaluationException> res = null;
+			try {
+				res = strategy.evaluate(query, queryBindings);
+
+				// mark the query as PassedThrough, such that outer result handlers are aware of this
+				// Note: for SingleSourceQuery (i.e. where we use pass through) res is explicitly
+				// EmptyIteration. Thus we can use it as indicator
+				if (originalQuery instanceof PassThroughTupleExpr && res instanceof EmptyIteration) {
+					((PassThroughTupleExpr) originalQuery).setPassedThrough(true);
+				}
+				res = new StopRemainingExecutionsOnCloseIteration(res, queryInfo);
+				return res;
+			} catch (Throwable t) {
+				if (res != null) {
+					res.close();
+				}
+				throw t;
```

Comment:
This is my approach to making sure that we handle any possible exceptions. In this particular case there might not be any possible exceptions, but I've previously come across constructors with null checks that would potentially fail without closing the iteration. The StopRemainingExecutionsOnCloseIteration constructor doesn't contain any null checks, but it might in the future.